### PR TITLE
3352 Sections with no `ConceptInstances` don't have "empty" text

### LINKED ISSUE #3352: Sections with no `ConceptInstance`s don't have "empty" text
Most sections have text that is displayed in the case that they are empty:
image

However, the following sections built from `ConceptInstance`s lack this text:

- Goals (see #3342)
- Assumptions
- Functional Requirements (technically there is always at least one, but this requirement might not have any inputs in its table, which is also a bug)
- Non-Functional Requirements
- Likely Changes
- Unlikely Changes

image
image
image

### PR DESCRIPTION
**Pull request by harmanpreet-sagar**
Closes #3352. Changes made to `drasil-docLang`, template example and stable. In case you are wondering why the branch name from which the PR is created is different than the issue number, it is because I began to solve #3429 but in the process resolved this issue.

### CODE REVIEW

**Review by smiths — CHANGES_REQUESTED** [review]
@harmanpreet-sagar I reviewed the changes to stable.  They look good to me.  However, they aren't entirely consistent.  Almost all of them are of the form "There is/are no X/Xs."  However, the following sections are inconsistent:

- Goal Statements
- Traceability Matrices and Graphs

Please make these two sections consistent with the others.

**Comment by harmanpreet-sagar** [comment]
Ok will fix those right now!

**Review by smiths — APPROVED** [review]
My requested changes were made.  The changes to stable look good.

File: code/drasil-docLang/lib/Drasil/Sections/TableOfAbbAndAcronyms.hs
Comment (samm82) [inline code/drasil-docLang/lib/Drasil/Sections/TableOfAbbAndAcronyms.hs:21]: Can this be represented using the pattern I mentioned before or is this completely different? As is, there is no trailing period like in the other examples; using a common helper function would reduce the number of potential errors like this:
image

File: code/drasil-docLang/lib/Drasil/Sections/Introduction.hs
Comment (samm82) [inline code/drasil-docLang/lib/Drasil/Sections/Introduction.hs:138]:
```suggestion
  S "do not need any prerequisite knowledge"]]
```

**Review by JacquesCarette — CHANGES_REQUESTED** [review]

File: code/drasil-docLang/lib/Drasil/DocumentLanguage/TraceabilityGraph.hs
Comment (smiths) [reply code/drasil-docLang/lib/Drasil/DocumentLanguage/TraceabilityGraph.hs:31]: Yes, "or" would be better.

File: code/drasil-docLang/lib/Drasil/Sections/SpecificSystemDescription.hs
Comment (smiths) [reply code/drasil-docLang/lib/Drasil/Sections/SpecificSystemDescription.hs:72]: I think "There is no terminology or definitions" would sound better.

File: code/drasil-docLang/lib/Drasil/Sections/Introduction.hs
Comment (harmanpreet-sagar) [reply code/drasil-docLang/lib/Drasil/Sections/Introduction.hs:138]: I will make that change right now!

File: code/drasil-docLang/lib/Drasil/Sections/SpecificSystemDescription.hs
Comment (harmanpreet-sagar) [reply code/drasil-docLang/lib/Drasil/Sections/SpecificSystemDescription.hs:72]: I will change that right now!

File: code/drasil-docLang/lib/Drasil/Sections/TableOfAbbAndAcronyms.hs
Comment (harmanpreet-sagar) [reply code/drasil-docLang/lib/Drasil/Sections/TableOfAbbAndAcronyms.hs:21]: Could you please clarify which pattern are you referring to @samm82?

Comment (harmanpreet-sagar) [reply code/drasil-docLang/lib/Drasil/Sections/TableOfAbbAndAcronyms.hs:21]: Oh ok, I don't think we can for this one because the pattern will convert the sentence to contents but the `Paragraph` section needs to have a `Sentence` as its input. However, I think to fix this, I just needed to change `+:+` to `+:+.`.

File: code/drasil-docLang/lib/Drasil/DocumentLanguage.hs
Comment (harmanpreet-sagar) [reply code/drasil-docLang/lib/Drasil/DocumentLanguage.hs:408]: I will fix that right now!

File: code/drasil-docLang/lib/Drasil/Sections/GeneralSystDesc.hs
Comment (harmanpreet-sagar) [reply code/drasil-docLang/lib/Drasil/Sections/GeneralSystDesc.hs:20]: Ok, I will create that helper right now!

Comment (harmanpreet-sagar) [reply code/drasil-docLang/lib/Drasil/Sections/GeneralSystDesc.hs:20]: Ok, I will fix that right now!

File: code/drasil-docLang/lib/Drasil/Sections/GeneralSystDesc.hs
Comment (samm82) [inline code/drasil-docLang/lib/Drasil/Sections/GeneralSystDesc.hs:27]: Might as well just inline this (e.g., remove the definition of `systCon_none`)

File: code/drasil-docLang/lib/Drasil/Sections/Requirements.hs
Comment (samm82) [inline code/drasil-docLang/lib/Drasil/Sections/Requirements.hs:98]: Nice catch here!

File: code/drasil-lang/lib/Language/Drasil/Sentence/Fold.hs
Comment (samm82) [inline code/drasil-lang/lib/Language/Drasil/Sentence/Fold.hs:133]: I feel like this isn't the best place for this. Is there somewhere in `drasil-utils` or `drasil-doclang` that would be a better fit? This doesn't really "fold" a `Sentence`.

**Review by samm82 — CHANGES_REQUESTED** [review]

File: code/drasil-lang/lib/Language/Drasil/Sentence/Fold.hs
Comment (harmanpreet-sagar) [reply code/drasil-lang/lib/Language/Drasil/Sentence/Fold.hs:133]: Ok, thank you for clarifying! After I submitted the PR, I was also wondering if that belongs there. I will make that change right now!

File: code/drasil-docLang/lib/Drasil/Sections/GeneralSystDesc.hs
Comment (harmanpreet-sagar) [reply code/drasil-docLang/lib/Drasil/Sections/GeneralSystDesc.hs:28]: I will make that change right now!

File: code/drasil-docLang/lib/Drasil/DocumentLanguage.hs
Comment (samm82) [reply code/drasil-docLang/lib/Drasil/DocumentLanguage.hs:409]: I think a way to do this would be to move the `[]` case of `introUCs`, etc. to `introChgs`

File: code/drasil-docLang/lib/Drasil/Sections/ReferenceMaterial.hs
Comment (samm82) [inline code/drasil-docLang/lib/Drasil/Sections/ReferenceMaterial.hs:16]: If this took a list of `NamedIdea`s, we would be able to use it when there are multiple (e.g., "There are no <x>s or <y>s.") using `foldlList Comma Options`.

**Review by samm82 — CHANGES_REQUESTED** [review]

**Comment by harmanpreet-sagar** [comment]
@samm82 @balacij I was trying to merge master into my branch and ran into the errors, which stated that the template file log was not empty. Am I supposed to be getting errors for this or not? I also got many warnings regarding unused imports. For reference, this is the log for the template that was generated.

[template_log.log](https://github.com/JacquesCarette/Drasil/files/11758947/template_log.log)

File: code/drasil-docLang/lib/Drasil/Sections/Introduction.hs
Comment (samm82) [reply code/drasil-docLang/lib/Drasil/Sections/Introduction.hs:68]: The helper that you created uses "There are no" instead of "There is no" - I wonder if there should be one helper for singular entries and one for plural ones.

**Comment by harmanpreet-sagar** [comment]
I did add chunks to the chunk database of the template. So do you think that I should just run `make stabilize` to solve the issue?

File: code/drasil-docLang/lib/Drasil/Sections/ReferenceMaterial.hs
Comment (harmanpreet-sagar) [reply code/drasil-docLang/lib/Drasil/Sections/ReferenceMaterial.hs:16]: Thats a good idea! I will change that right now!

File: code/drasil-docLang/lib/Drasil/Sections/SpecificSystemDescription.hs
Comment (harmanpreet-sagar) [reply code/drasil-docLang/lib/Drasil/Sections/SpecificSystemDescription.hs]: Could I change it to either "There is no terminology or definition" or "There are no terminologies or definitions"? As doing so would help use the helper functions that we have created such as `emptySectSentSing` or `emptySectSentPlu` respectively.

Comment (samm82) [reply code/drasil-docLang/lib/Drasil/Sections/SpecificSystemDescription.hs]: I would change it to "There are no terms or definitions." - A _terminology_ is a collection of words with specific meanings, while a _term_ is just one of these words. We have one _terminology_ but multiple _terms_.

**Comment by samm82** [comment]
I would look at the contents of the logs to make sure that all of the discrepancies are expected before running `make stabilize`

**Comment by harmanpreet-sagar** [comment]
Sounds good!

File: code/drasil-docLang/lib/Drasil/Sections/ReferenceMaterial.hs
Comment (JacquesCarette) [inline code/drasil-docLang/lib/Drasil/Sections/ReferenceMaterial.hs:14]: Very nice!

**Review by JacquesCarette — APPROVED** [review]

**Comment by JacquesCarette** [comment]
All up to @samm82 to double-check (and then go ahead and merge since @smiths and I have both approved)

**Comment by samm82** [comment]
I had a "minor" tweak that I wanted to make that took longer than anticipated - I don't have time to finish it up right now, but I'll try to have it for tomorrow!

**Review by samm82 — COMMENTED** [review]
I've made some changes now; while I (semi-obviously) approve of them, it feels weird just merging it in, so I'm going to wait for @harmanpreet-sagar's approval (make sure to understand why I made the changes I did, and reach out if you have questions!)

**Comment by harmanpreet-sagar** [comment]
@samm82 I like them! A much more efficient way to carry out that task! Thanks!! 😄

**Review by samm82 — APPROVED** [review]

**Pull request merged**
